### ISSUE TITLE
amp could be able to encode DateTime and Decimal

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2651]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)| @bitglue reported|
|-|-|
|Trac ID|trac#2651|
|Type|enhancement|
|Created|2007-05-15 19:44:40Z|
|Branch|https://github.com/twisted/twisted/tree/amp-datetime-decimal-2651|

Dates and times are cool. Decimals are great.

Attachments:

* [ampext.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/93a/93aeee8c1ad46e038c3e3ff8461b9a914eb84cdb/5ce4ffe53c3b6245b734389d913cf31aa27a7702.py) (4138 bytes) - added by indigo on 2007-05-15 19:50:17Z - implementation of nice things
* [2651.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/93a/93aeee8c1ad46e038c3e3ff8461b9a914eb84cdb/3799edbdbdd1222f698caafa13c8d371c97608d8.patch) (9551 bytes) - added by thomasvs on 2010-03-21 15:54:16Z - reworked patch, integrating into amp.py, with unit tests

<details><summary>Searchable metadata</summary>

```
trac-id__2651 2651
type__enhancement enhancement
reporter__indigo indigo
priority__normal normal
milestone__ 
branch__branches_amp_datetime_decimal_2651 branches/amp-datetime-decimal-2651
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__amp amp
time__1179258280000000 1179258280000000
changetime__1273158420000000 1273158420000000
version__None None
owner__ 
cc__therve cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

That would be even better if tests where unit tests.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to @bitglue|
|-|-|

Thanks!

But yeah, as therve says, can you convert the test to a unit test and the implementation into a unified diff?  [This page has some useful information](https://twistedmatrix.com/trac/wiki/TwistedDevelopment) about requirements for integrating patches, etc.

Not all of this should be going into one file.  The fixed-offset timezone, for example, should probably live in twisted.python somewhere.  For that matter, maybe the right way to implement that is to bring across the excellent ```Time``` class you did for Divmod and put it into twisted.python somewhere and use that for representing time objects on the wire.

**Comment by twisted-trac** [comment]
|<img alt="thomasvs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|thomasvs removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [28860]) Branching to 'amp-datetime-decimal-2651'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to thomasvs|
|-|-|

Looks good! I have only one comment:

 * Most test classes and test methods need a docstring.

Please provide the patch against the branch I made if psossible.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [28862]) Address some doc and testing issues

refs [#2651](https://github.com/twisted/twisted/issues/2651)

  * Clarify the docstrings for several of the ListOf combo test
  * Rename the ListOf Decimal NaN test case to follow the convention
  * Add docstrings for classes and methods which were missing them
  * Convert the Decimal NaN test to use assertTrue instead of failUnless
  * Add a test case which directly tests amp.Decimal (the other always also tested ListOf)
  * Rename amp.UTCtzinfo to amp.utc
  * Add some more tests for amp._FixedOffsetTZInfo
  * Add some specification for the format used by amp.Decimal
  * Reject all non-decimal.Decimal instances in amp.Decimal.toString
  * Fix exception raising style in amp.DateTime.fromString
  * Tweak the exception raised when a naive datetime is encountered

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [28863]) Add some more unit tests for datetime and... simplify? the implementation a bit

refs [#2651](https://github.com/twisted/twisted/issues/2651)

  * Add a reference to [#4409](https://github.com/twisted/twisted/issues/4409) near the new XXX
  * Add direct unit tests for DateTime.toString and .fromString that should exercise all of its parsing code
  * Split the _FixedOffsetTZInfo tests off into another testcase again
  * Add some more docs for _FixedOffsetTZInfo
  * Simplify the DateTime parser to just consider offsets, not try to split up the string based on values

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [28864]) News fragment

refs [#2651](https://github.com/twisted/twisted/issues/2651)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Thanks for the review Thomas.  I started off just looking at docstring improvements, but I got caught up in things. :)

All my changes are linked in the three comments above.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Ah, also, [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/amp-datetime-decimal-2651).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [28865]) Stop using Decimal.is_{qnan,snan,signed}; they're new in 2.6

refs [#2651](https://github.com/twisted/twisted/issues/2651)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj set owner to @exarkun<br>@wallrj set status to `new`|
|-|-|

 * Tests pass
 
## [branches/amp-datetime-decimal-2651/twisted/test/test_amp.py](branches/amp-datetime-decimal-2651/twisted/test/test_amp.py)
 0. Remove comment about missing test body
```
1484	        # XXX something is missing.  See #4409. 
```

## [branches/amp-datetime-decimal-2651/twisted/protocols/amp.py](branches/amp-datetime-decimal-2651/twisted/protocols/amp.py)
 0. I may have misunderstood the NaN stuff, but surely Quiet NaN would be encoded as qNaN - in which case the NaN tests may need updating
```
	2458	      - Quiet not-a-number is encoded as either C{"NaN"} or C{"-NaN"} 
 	2459	      - Signalling not-a-number is encoded as either C{"qNaN"} or C{"-qNaN"} 
```

Otherwise looks good.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [28904]) signalling nan is sNaN not qNaN

refs [#2651](https://github.com/twisted/twisted/issues/2651)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [28905]) maybe this won't conflict if we get rid of the XXX

refs [#2651](https://github.com/twisted/twisted/issues/2651)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [28906]) Merge amp-datetime-decimal-2651

Author: indigo, thomasvs, Zaheer Merali, exarkun
Reviewer: therve, glyph, exarkun, rwall
Fixes: [#2651](https://github.com/twisted/twisted/issues/2651)

Add AMP argument types for serializing timestamps (Python datetime instances)
and decimals (Python Decimal instances).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
